shared-bindings/bitmaptools/__init__.c

```diff
@@ -0,0 +1,254 @@
+//|        source_bitmap: displayio.Bitmap, px: int, py: int, source_clip0: Tuple[int, int], source_clip1: Tuple[int, int],
+//|        angle: float, scale: float, skip_index: int) -> None:
+//|      """Inserts the source bitmap region into the destination bitmap with rotation
+//|      (angle), scale and clipping (both on source and destination bitmaps).
+//|
+//|      :param bitmap dest_bitmap: The bitmap that will be copied into
+//|      :param int ox: Horizontal pixel location in destination bitmap where source bitmap
+//|             point (px,py) is placed
+//|      :param int oy: Vertical pixel location in destination bitmap where source bitmap
+//|             point (px,py) is placed
+//|      :param Tuple[int,int] dest_clip0: First corner of rectangular destination clipping
+//|             region that constrains region of writing into destination bitmap
+//|      :param Tuple[int,int] dest_clip1: Second corner of rectangular destination clipping
+//|             region that constrains region of writing into destination bitmap
+//|      :param bitmap source_bitmap: Source bitmap that contains the graphical region to be copied
+//|      :param int px: Horizontal pixel location in source bitmap that is placed into the
+//|             destination bitmap at (ox,oy)
+//|      :param int py: Vertical pixel location in source bitmap that is placed into the
+//|             destination bitmap at (ox,oy)
+//|      :param Tuple[int,int] source_clip0: First corner of rectangular source clipping
+//|             region that constrains region of reading from the source bitmap
+//|      :param Tuple[int,int] source_clip1: Second corner of rectangular source clipping
+//|             region that constrains region of reading from the source bitmap
+//|      :param float angle: Angle of rotation, in radians (positive is clockwise direction)
+//|      :param float scale: Scaling factor
+//|      :param int skip_index: Bitmap palette index in the source that will not be copied,
+//|             set to None to copy all pixels"""
+//|      ...
+//|
+STATIC mp_obj_t bitmaptools_obj_rotozoom(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args){
+    enum {ARG_dest_bitmap, ARG_ox, ARG_oy, ARG_dest_clip0, ARG_dest_clip1,
+        ARG_source_bitmap, ARG_px, ARG_py,
+        ARG_source_clip0, ARG_source_clip1,
+        ARG_angle, ARG_scale, ARG_skip_index};
+
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_dest_bitmap, MP_ARG_REQUIRED | MP_ARG_OBJ},
+        {MP_QSTR_ox, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} }, // None convert to destination->width  / 2
+        {MP_QSTR_oy, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} }, // None convert to destination->height / 2
+        {MP_QSTR_dest_clip0, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        {MP_QSTR_dest_clip1, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+
+        {MP_QSTR_source_bitmap, MP_ARG_REQUIRED | MP_ARG_OBJ},
```

Comment:
I know this order makes sense here but in practice all required params should be first in the list here and above in the docs. They need to go before a `*,` which indicates where the kwarg-only parameters start in the function declaration.

Reply:
Thanks for the review and for the feedback.  I actually struggled with understanding how to use the `MP_DEFINE_CONST_FUN_OBJ_KW`.  I pulled this out of Bitmap.c and into its own file, I went from use of `my_bitmap.rotozoom(...)` to `rotozoom(my_bitmap,...)`.

But I struggled for a while since I forgot to update the `mp_arg_parse_all` to not subtract an argument! 

I updated per your request and tested this with positional-style usage, and with parameter-name usage and both worked fine.

Thank you for the prompt feedback!